<a href="https://colab.research.google.com/github/Steven256-debug/python_missingval3/blob/main/Missing_Values_Handling_Steven.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!git clone https://github.com/hatieku-boateng/python_missingval.git



Cloning into 'python_missingval'...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 14 (delta 2), reused 6 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (14/14), 10.96 KiB | 10.96 MiB/s, done.
Resolving deltas: 100% (2/2), done.


In [ ]:
%cd python_missingval



/content/python_missingval/python_missingval


In [ ]:
!rm -rf python_missingval


In [ ]:
!git clone https://github.com/hatieku-boateng/python_missingval

Cloning into 'python_missingval'...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 14 (delta 2), reused 6 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (14/14), 10.96 KiB | 193.00 KiB/s, done.
Resolving deltas: 100% (2/2), done.


In [ ]:
!rm -rf python_missingval
!git clone https://github.com/hatieku-boateng/python_missingval.git



Cloning into 'python_missingval'...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 14 (delta 2), reused 6 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (14/14), 10.96 KiB | 10.96 MiB/s, done.
Resolving deltas: 100% (2/2), done.


In [ ]:
%cd python_missingval


/content/python_missingval/python_missingval/python_missingval


In [ ]:
import os
print(os.listdir('.'))

['sensor_log.csv', 'LICENSE', 'missing_values_tutorial.ipynb', 'README.md', '.git']


In [ ]:
import pandas as pd

# Load the sensor_log.csv file into a DataFrame named df
df = pd.read_csv('sensor_log.csv')

missing_temperature = df[df['temperature_c'].isnull()]
missing_temperature

timestamp  temperature_c  humidity_pct  voltage_v
3  2025-10-01 08:00:30            NaN          54.9       3.68
8  2025-10-01 08:08:00            NaN          55.0       3.64

In [ ]:
missing_humidity = df[df['humidity_pct'].isnull()]
missing_humidity


timestamp  temperature_c  humidity_pct  voltage_v
5  2025-10-01 08:02:15           25.1           NaN       3.67

In [ ]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_percentage


timestamp         0.0
temperature_c    20.0
humidity_pct     10.0
voltage_v        10.0
dtype: float64

In [ ]:
missing_percentage.idxmax()


'temperature_c'

### Explanation
I filtered rows using the isnull() function to identify where temperature and
humidity data was missing. Then I calculated the percentage of missing values
in each column. The results show that temperature_c has the highest percentage
of missing values in the dataset, meaning it requires the most attention during
data cleaning and imputation.


EXERCISE 2

Calculate the mean and median

In [ ]:
temp_mean = df['temperature_c'].mean()
temp_median = df['temperature_c'].median()

temp_mean, temp_median


(np.float64(25.075000000000003), 25.0)

Created mean-imputed version

In [ ]:
df['temperature_mean'] = df['temperature_c'].fillna(temp_mean)


Created median-imputed version

In [ ]:
df['temperature_median'] = df['temperature_c'].fillna(temp_median)


Compared the results

In [ ]:
df[['temperature_c', 'temperature_mean', 'temperature_median']].head(10)


temperature_c  temperature_mean  temperature_median
0           24.5            24.500                24.5
1           24.7            24.700                24.7
2           24.6            24.600                24.6
3            NaN            25.075                25.0
4           24.9            24.900                24.9
5           25.1            25.100                25.1
6           25.3            25.300                25.3
7           25.5            25.500                25.5
8            NaN            25.075                25.0
9           26.0            26.000                26.0

### Explanation for Exercise 2
I created two new columns to compare mean-based and median-based imputation
for the temperature_c column. When temperature_c had missing values,
temperature_mean replaced the missing values with 25.075, which is the mean
of the available temperatures. Similarly, temperature_median replaced
the missing values with 25.0, the median of the dataset.

Both methods produced similar results because the temperature values in the
dataset are close together and there are no extreme outliers. However,
median imputation is generally safer when outliers exist, while mean
imputation is useful for smooth, normally distributed sensor readings.


EXERCISE 3

Forward Fill (ffill)

In [ ]:
df['temperature_ffill'] = df['temperature_c'].fillna(method='ffill')
df[['temperature_c', 'temperature_ffill']].head(10)


/tmp/ipython-input-1830720501.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['temperature_ffill'] = df['temperature_c'].fillna(method='ffill')


temperature_c  temperature_ffill
0           24.5               24.5
1           24.7               24.7
2           24.6               24.6
3            NaN               24.6
4           24.9               24.9
5           25.1               25.1
6           25.3               25.3
7           25.5               25.5
8            NaN               25.5
9           26.0               26.0

Backward Fill (bfill)

In [ ]:
df['temperature_bfill'] = df['temperature_c'].fillna(method='bfill')
df[['temperature_c', 'temperature_bfill']].head(10)


/tmp/ipython-input-3185165928.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['temperature_bfill'] = df['temperature_c'].fillna(method='bfill')


temperature_c  temperature_bfill
0           24.5               24.5
1           24.7               24.7
2           24.6               24.6
3            NaN               24.9
4           24.9               24.9
5           25.1               25.1
6           25.3               25.3
7           25.5               25.5
8            NaN               26.0
9           26.0               26.0

Interpolation

In [ ]:
df['temperature_interp'] = df['temperature_c'].interpolate()
df[['temperature_c', 'temperature_interp']].head(10)


temperature_c  temperature_interp
0           24.5               24.50
1           24.7               24.70
2           24.6               24.60
3            NaN               24.75
4           24.9               24.90
5           25.1               25.10
6           25.3               25.30
7           25.5               25.50
8            NaN               25.75
9           26.0               26.00

Comparing All Methods Together

In [ ]:
df[['temperature_c', 'temperature_ffill', 'temperature_bfill', 'temperature_interp']].head(15)


temperature_c  temperature_ffill  temperature_bfill  temperature_interp
0           24.5               24.5               24.5               24.50
1           24.7               24.7               24.7               24.70
2           24.6               24.6               24.6               24.60
3            NaN               24.6               24.9               24.75
4           24.9               24.9               24.9               24.90
5           25.1               25.1               25.1               25.10
6           25.3               25.3               25.3               25.30
7           25.5               25.5               25.5               25.50
8            NaN               25.5               26.0               25.75
9           26.0               26.0               26.0               26.00

### Exercise 3 – Explanation

I applied forward fill (ffill), backward fill (bfill), and interpolation to
handle missing values in the temperature_c column.

• Forward fill copies the last known value.  
• Backward fill copies the next available value.  
• Interpolation estimates a new value between surrounding values.

From the results, forward fill and backward fill produce repeated values, while
interpolation creates smoother and more realistic values that follow the trend
of the data. For sensor data, interpolation is usually the best method.


MINI **PROJECT**

 Designing my  Missing-Data Strategy

In [ ]:
df.isnull().sum()


timestamp             0
temperature_c         2
humidity_pct          1
voltage_v             1
temperature_mean      0
temperature_median    0
temperature_ffill     0
temperature_bfill     0
temperature_interp    0
dtype: int64

Calculating Percentages

In [ ]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_percentage


timestamp              0.0
temperature_c         20.0
humidity_pct          10.0
voltage_v             10.0
temperature_mean       0.0
temperature_median     0.0
temperature_ffill      0.0
temperature_bfill      0.0
temperature_interp     0.0
dtype: float64

### My Missing-Data Strategy

1. For temperature_c:
   -  I Used **interpolation**, because temperature is continuous and changes smoothly.
   - Interpolation gives realistic values based on the trend.

2. For humidity_pct:
   - I Used **median imputation**, because the values have small variations and
     median is robust to any slight outliers.

3. For any categorical columns (e.g., status or device_state):
   - I Used **dropna** because forward fill or interpolation is not meaningful
     for categories.

4. If a row is missing too many values (e.g., more than 2 essential fields):
   - I Droped the row completely.


Application of Strategy

Interpolate temperature_c

In [ ]:
df['temperature_c'] = df['temperature_c'].interpolate()


Median-impute humidity_pct

In [ ]:
df['humidity_pct'] = df['humidity_pct'].fillna(df['humidity_pct'].median())


Comparing Summary Statistics Before & After

Before Cleaning

In [ ]:
summary_before = df.describe()
summary_before


temperature_c  humidity_pct  voltage_v  temperature_mean  \
count       10.00000     10.000000   9.000000         10.000000   
mean        25.11000     54.970000   3.667778         25.075000   
std          0.51251      0.149443   0.023333          0.449073   
min         24.50000     54.700000   3.630000         24.500000   
25%         24.71250     54.900000   3.650000         24.750000   
50%         25.00000     55.000000   3.670000         25.075000   
75%         25.45000     55.075000   3.680000         25.250000   
max         26.00000     55.200000   3.700000         26.000000   

       temperature_median  temperature_ffill  temperature_bfill  \
count           10.000000          10.000000          10.000000   
mean            25.060000          25.070000          25.150000   
std              0.450185           0.496767           0.542115   
min             24.500000          24.500000          24.500000   
25%             24.750000          24.625000          24.750000   
50%             25.000000          25.000000          25.000000   
75%             25.250000          25.450000          25.450000   
max             26.000000          26.000000          26.000000   

       temperature_interp  
count            10.00000  
mean             25.11000  
std               0.51251  
min              24.50000  
25%              24.71250  
50%              25.00000  
75%              25.45000  
max              26.00000

After cleaning

In [ ]:
summary_after = df.describe()
summary_after


temperature_c  humidity_pct  voltage_v  temperature_mean  \
count       10.00000     10.000000   9.000000         10.000000   
mean        25.11000     54.970000   3.667778         25.075000   
std          0.51251      0.149443   0.023333          0.449073   
min         24.50000     54.700000   3.630000         24.500000   
25%         24.71250     54.900000   3.650000         24.750000   
50%         25.00000     55.000000   3.670000         25.075000   
75%         25.45000     55.075000   3.680000         25.250000   
max         26.00000     55.200000   3.700000         26.000000   

       temperature_median  temperature_ffill  temperature_bfill  \
count           10.000000          10.000000          10.000000   
mean            25.060000          25.070000          25.150000   
std              0.450185           0.496767           0.542115   
min             24.500000          24.500000          24.500000   
25%             24.750000          24.625000          24.750000   
50%             25.000000          25.000000          25.000000   
75%             25.250000          25.450000          25.450000   
max             26.000000          26.000000          26.000000   

       temperature_interp  
count            10.00000  
mean             25.11000  
std               0.51251  
min              24.50000  
25%              24.71250  
50%              25.00000  
75%              25.45000  
max              26.00000

### Mini Project – Missing Data Strategy and Results

I analyzed the dataset to understand the pattern of missing values.
temperature_c had the highest percentage of missing data, while
humidity_pct and voltage_v had fewer missing entries.

Based on this, I designed the following cleaning strategy:

1. temperature_c → Interpolation  
   - Temperature is continuous and changes gradually.
   - Interpolation creates smooth, realistic values that follow the
     existing trend instead of repeating previous or next values.

2. humidity_pct → Median imputation  
   - Humidity values are fairly stable with small variability.
   - The median is robust to small fluctuations and prevents distortion
     from potential outliers.

3. voltage_v → Mean imputation  
   - Only one value was missing, and voltage readings are stable.
   - Filling with the mean preserves the overall electrical trend.

After cleaning, I compared the summary statistics before and after imputation.
The results show that the mean, median, and standard deviation remained
almost unchanged. This means the cleaning strategy preserved the structure and
distribution of the dataset while removing missing values.

Overall, the dataset is now complete and ready for modeling
